algorithms/src/std_algorithms/Kokkos_NonModifyingSequenceOperations.hpp

```diff
@@ -0,0 +1,2100 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_NON_MODIFYING_SEQUENCE_OPERATIONS_HPP
+#define KOKKOS_NON_MODIFYING_SEQUENCE_OPERATIONS_HPP
+
+#include <Kokkos_Core.hpp>
+#include "Kokkos_BeginEnd.hpp"
+#include "Kokkos_Constraints.hpp"
+#include "Kokkos_MinMaxOperations.hpp"
+#include "Kokkos_ModifyingOperations.hpp"
+#include "Kokkos_HelperPredicates.hpp"
+#include "Kokkos_RandomAccessIterator.hpp"
+#include <string>
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+
+// ------------------------------------------
+//
+// functors
+//
+// ------------------------------------------
+
+template <bool is_find_if, class IndexType, class IteratorType,
```

Comment:
Why is `is_find_if` at the beginning?  If we can put it at the end, we can default it to `true`.

We could also add:

```
template<class IndexType, class IteratorType, class ReducerType, class PredicateType>
using StdFindIfFunctor = StdFindIfOrNotFunctor<IndexType, IteratorType, ReducerType, PredicateType, true>;
```

Update: I see why it is first (in `find_if_or_not_impl`) and why we don't necessarily need deduction guides/functions.

Reply:
ok thanks !